File: Tools/python/simple_bot.py
from __future__ import annotations

import argparse
from typing import Any, Dict, Optional

from game_bridge import GameBridgeClient


def choose_action(snapshot: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    scene_type = snapshot.get("sceneType")
    if scene_type == "main_menu":
        return {"kind": "start_new_run"}
    if scene_type == "map":
        return choose_map_action(snapshot.get("map") or {})
    if scene_type == "battle":
        return choose_battle_action(snapshot.get("battle") or {})
    if scene_type == "reward":
        return choose_reward_action(snapshot.get("reward") or {})
    if scene_type == "event":
        return {"kind": "choose_event_option", "optionIndex": 0}
    return choose_fallback_action(snapshot)


def choose_map_action(map_state: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    rows = map_state.get("rows") or []
    current_row = map_state.get("currentRow")
    for row in rows:
        if row.get("rowIndex") != current_row:
            continue

        candidates = [node for node in row.get("nodes", []) if node.get("canSelect")]
        if not candidates:
            return None

        priority = {
            "NormalBattle": 0,
            "EliteBattle": 1,
            "Event": 2,
            "Rest": 3,
            "Shop": 4,
        }
        node = sorted(candidates, key=lambda item: (priority.get(item.get("nodeType"), 99), item.get("column", 99)))[0]
        return {"kind": "choose_map_node", "column": node.get("column")}

    return None


def choose_battle_action(battle: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    if battle.get("battleEnded"):
        return None

    hand = [card for card in battle.get("hand", []) if card.get("isPlayable")]
    enemies = [enemy for enemy in battle.get("enemies", []) if enemy.get("isAlive")]
    if not hand:
        return {"kind": "end_turn"}

    lethal = try_find_lethal(hand, enemies)
    if lethal is not None:
        return lethal

    debuff = next((card for card in hand if has_any(card, "Vulnerable", "易伤")), None)
    if debuff is not None and enemies:
        target = sorted(enemies, key=lambda enemy: enemy.get("hp", 0) + enemy.get("block", 0), reverse=True)[0]
        return {"kind": "play_card", "handIndex": debuff.get("handIndex"), "targetEnemyIndex": target.get("enemyIndex")}

    attack = next((card for card in sorted(hand, key=estimate_damage, reverse=True) if estimate_damage(card) > 0), None)
    if attack is not None:
        if attack.get("requiresEnemyTarget") and enemies:
            target = sorted(enemies, key=lambda enemy: enemy.get("hp", 0) + enemy.get("block", 0))[0]
            return {"kind": "play_card", "handIndex": attack.get("handIndex"), "targetEnemyIndex": target.get("enemyIndex")}
        return {"kind": "play_card", "handIndex": attack.get("handIndex")}

    defense = next((card for card in hand if has_any(card, "Block", "格挡")), None)
    if defense is not None:
        return {"kind": "play_card", "handIndex": defense.get("handIndex")}

    return {"kind": "end_turn"}


def try_find_lethal(hand: list[Dict[str, Any]], enemies: list[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    for enemy in sorted(enemies, key=lambda item: item.get("hp", 0) + item.get("block", 0)):
        effective_hp = enemy.get("hp", 0) + enemy.get("block", 0)
        for card in sorted(hand, key=estimate_damage, reverse=True):
            if not card.get("requiresEnemyTarget"):
                continue
            if estimate_damage(card) >= effective_hp:
                return {"kind": "play_card", "handIndex": card.get("handIndex"), "targetEnemyIndex": enemy.get("enemyIndex")}
    return None


def choose_reward_action(reward: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    if reward.get("mode") == "card_pack":
        options = reward.get("cardOptions") or []
        if not options:
            return {"kind": "skip_reward"}

        best = sorted(options, key=score_reward_option, reverse=True)[0]
        return {"kind": "choose_reward_card", "optionIndex": best.get("optionIndex")}

    reward_types = reward.get("rewardTypes") or []
    for reward_type in ("relic", "card_pack", "potion", "random"):
        if reward_type in reward_types:
            return {"kind": "choose_reward_type", "rewardType": reward_type}

    return {"kind": "skip_reward"}


def choose_fallback_action(snapshot: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    legal_actions = snapshot.get("legalActions") or []
    if not legal_actions:
        return None

    first = legal_actions[0]
    action = {"kind": first.get("kind")}
    parameters = first.get("parameters") or {}

    for key in ("column", "handIndex", "cardId", "rewardType", "optionIndex"):
        if key in parameters:
            action[key] = parameters[key]

    target_enemy_indices = parameters.get("targetEnemyIndices") or []
    if target_enemy_indices:
        action["targetEnemyIndex"] = target_enemy_indices[0]

    return action


def score_reward_option(option: Dict[str, Any]) -> int:
    text = f"{option.get('name', '')} {option.get('description', '')}"
    score = 0
    if has_text(text, "Strength", "力量"):
        score += 40
    if has_text(text, "Energy", "能量"):
        score += 35
    if has_text(text, "Vulnerable", "易伤"):
        score += 20
    score += sum(1 for ch in text if ch.isdigit())
    return score


def estimate_damage(card: Dict[str, Any]) -> int:
    text = card.get("description", "")
    best = 0
    current = []
    for ch in text:
        if ch.isdigit():
            current.append(ch)
            continue

        if current:
            best = max(best, int("".join(current)))
            current = []

    if current:
        best = max(best, int("".join(current)))
    return best


def has_any(card: Dict[str, Any], *keywords: str) -> bool:
    text = card.get("description", "")
    return has_text(text, *keywords)


def has_text(text: str, *keywords: str) -> bool:
    return any(keyword.lower() in text.lower() for keyword in keywords)


def main() -> int:
    parser = argparse.ArgumentParser(description="Simple rule-based bot for Slay the HS.")
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=47077)
    parser.add_argument("--steps", type=int, default=500)
    args = parser.parse_args()

    client = GameBridgeClient(host=args.host, port=args.port)
    result = client.get_snapshot()
    snapshot = result.get("snapshot")
    if not snapshot:
        raise RuntimeError("The game bridge did not return a snapshot.")

    for step in range(args.steps):
        log_snapshot(step, snapshot)
        action = choose_action(snapshot)
        if action is None:
            print(f"bot stop: no action for scene {snapshot.get('sceneType')}")
            log_legal_actions(snapshot)
            return 0

        print(f"bot action: {format_action(action)}")
        response = client.execute_action(
            action["kind"],
            expected_state_version=snapshot.get("stateVersion"),
            **{k: v for k, v in action.items() if k != "kind"},
        )
        if not response.get("ok"):
            if "State version mismatch" in str(response.get("message")) and response.get("snapshot"):
                print("bot info: version mismatch, refresh snapshot and continue")
                snapshot = response["snapshot"]
                continue
            log_legal_actions(snapshot)
            raise RuntimeError(f"Action failed: {response.get('message')}")

        snapshot = response.get("snapshot")
        if not snapshot:
            raise RuntimeError("Missing snapshot after action.")

    print("bot stop: reached step limit")
    return 0


def log_snapshot(step: int, snapshot: Dict[str, Any]) -> None:
    run = snapshot.get("run") or {}
    legal_actions = snapshot.get("legalActions") or []
    print(
        "bot snapshot:",
        f"step={step}",
        f"scene={snapshot.get('sceneType')}",
        f"version={snapshot.get('stateVersion')}",
        f"floor={run.get('floor')}",
        f"hp={run.get('playerHp')}/{run.get('maxHp')}",
        f"wins={run.get('battlesWon')}",
        f"legal={len(legal_actions)}",
    )

    scene_type = snapshot.get("sceneType")
    if scene_type == "battle":
        battle = snapshot.get("battle") or {}
        alive = sum(1 for enemy in battle.get("enemies", []) if enemy.get("isAlive"))
        print(
            "bot battle:",
            f"turn={battle.get('turn')}",
            f"energy={battle.get('energy')}",
            f"hand={len(battle.get('hand', []))}",
            f"enemiesAlive={alive}",
        )
    elif scene_type == "reward":
        reward = snapshot.get("reward") or {}
        print("bot reward:", f"mode={reward.get('mode')}", f"types={reward.get('rewardTypes')}")
    elif scene_type == "event":
        event_state = snapshot.get("event") or {}
        print("bot event:", f"id={event_state.get('eventId')}", f"title={event_state.get('title')}")
    elif scene_type == "map":
        map_state = snapshot.get("map") or {}
        print("bot map:", f"row={map_state.get('currentRow')}", f"column={map_state.get('currentColumn')}")


def log_legal_actions(snapshot: Dict[str, Any]) -> None:
    legal_actions = snapshot.get("legalActions") or []
    if not legal_actions:
        print("bot legal: none")
        return

    for action in legal_actions[:8]:
        print(f"bot legal: {action.get('kind')} ({action.get('label')})")


def format_action(action: Dict[str, Any]) -> str:
    parts = [str(action.get("kind"))]
    for key in ("handIndex", "cardId", "targetEnemyIndex", "column", "rewardType", "optionIndex", "eventOption"):
        if key in action and action[key] is not None:
            parts.append(f"{key}={action[key]}")
    return " ".join(parts)


if __name__ == "__main__":
    raise SystemExit(main())

File: Tools/SlayHs.Agent/SimpleBot.cs
namespace SlayHs.Agent;

internal sealed class SimpleBot
{
    private readonly GameBridgeClient _client;

    public SimpleBot(GameBridgeClient client)
    {
        _client = client;
    }

    public async Task<int> RunAsync(CancellationToken cancellationToken)
    {
        Console.Error.WriteLine("simple-bot: waiting for game bridge...");

        SnapshotEnvelope snapshot;
        try
        {
            snapshot = (await _client.GetSnapshotAsync(cancellationToken)).Snapshot
                ?? throw new InvalidOperationException("Missing snapshot.");
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"simple-bot: cannot reach game bridge: {ex.Message}");
            return 1;
        }

        for (var step = 0; step < 500 && !cancellationToken.IsCancellationRequested; step++)
        {
            LogSnapshot(step, snapshot);
            var action = ChooseAction(snapshot);
            if (action == null)
            {
                Console.Error.WriteLine($"simple-bot: no action for scene '{snapshot.SceneType}', stopping.");
                LogLegalActions(snapshot);
                return 0;
            }

            Console.Error.WriteLine($"simple-bot: action -> {FormatAction(action)}");
            var result = await _client.ExecuteAsync(action, snapshot.StateVersion, cancellationToken);
            if (!result.Ok)
            {
                if (result.Message.Contains("State version mismatch", StringComparison.OrdinalIgnoreCase)
                    && result.Snapshot != null)
                {
                    Console.Error.WriteLine("simple-bot: version mismatch, refreshing snapshot and continuing.");
                    snapshot = result.Snapshot;
                    continue;
                }

                Console.Error.WriteLine($"simple-bot: action failed: {result.Message}");
                LogLegalActions(snapshot);
                return 1;
            }

            snapshot = result.Snapshot ?? throw new InvalidOperationException("Missing post-action snapshot.");

            if (snapshot.SceneType == "main_menu" && step > 0)
            {
                Console.Error.WriteLine("simple-bot: back on main menu, stopping.");
                return 0;
            }
        }

        Console.Error.WriteLine("simple-bot: reached step limit.");
        return 0;
    }

    private static void LogSnapshot(int step, SnapshotEnvelope snapshot)
    {
        var run = snapshot.Run;
        Console.Error.WriteLine(
            $"simple-bot: step={step} scene={snapshot.SceneType} version={snapshot.StateVersion} " +
            $"floor={run.Floor} hp={run.PlayerHp}/{run.MaxHp} wins={run.BattlesWon} legal={snapshot.LegalActions.Count}");

        switch (snapshot.SceneType)
        {
            case "battle" when snapshot.Battle != null:
                var battle = snapshot.Battle;
                var alive = battle.Enemies.Count(enemy => enemy.IsAlive);
                Console.Error.WriteLine(
                    $"simple-bot: battle turn={battle.Turn} energy={battle.Energy} hand={battle.Hand.Count} enemiesAlive={alive}");
                break;
            case "reward" when snapshot.Reward != null:
                Console.Error.WriteLine(
                    $"simple-bot: reward mode={snapshot.Reward.Mode} rewardTypes={string.Join(",", snapshot.Reward.RewardTypes)}");
                break;
            case "event" when snapshot.Event != null:
                Console.Error.WriteLine(
                    $"simple-bot: event id={snapshot.Event.EventId} title={snapshot.Event.Title}");
                break;
            case "map" when snapshot.Map != null:
                Console.Error.WriteLine(
                    $"simple-bot: map row={snapshot.Map.CurrentRow}");
                break;
        }
    }

    private static void LogLegalActions(SnapshotEnvelope snapshot)
    {
        if (snapshot.LegalActions.Count == 0)
        {
            Console.Error.WriteLine("simple-bot: legal actions -> none");
            return;
        }

        foreach (var action in snapshot.LegalActions.Take(8))
        {
            Console.Error.WriteLine($"simple-bot: legal -> {action.Kind} ({action.Label})");
        }
    }

    private static string FormatAction(ActionEnvelope action)
    {
        var parts = new List<string> { action.Kind };
        if (action.HandIndex.HasValue)
        {
            parts.Add($"handIndex={action.HandIndex.Value}");
        }
        if (!string.IsNullOrWhiteSpace(action.CardId))
        {
            parts.Add($"cardId={action.CardId}");
        }
        if (action.TargetEnemyIndex.HasValue)
        {
            parts.Add($"targetEnemyIndex={action.TargetEnemyIndex.Value}");
        }
        if (action.Column.HasValue)
        {
            parts.Add($"column={action.Column.Value}");
        }
        if (!string.IsNullOrWhiteSpace(action.RewardType))
        {
            parts.Add($"rewardType={action.RewardType}");
        }
        if (action.OptionIndex.HasValue)
        {
            parts.Add($"optionIndex={action.OptionIndex.Value}");
        }
        if (!string.IsNullOrWhiteSpace(action.EventOption))
        {
            parts.Add($"eventOption={action.EventOption}");
        }

        return string.Join(" ", parts);
    }

    private static ActionEnvelope? ChooseAction(SnapshotEnvelope snapshot)
    {
        var action = snapshot.SceneType switch
        {
            "main_menu" => new ActionEnvelope { Kind = "start_new_run" },
            "map" => ChooseMapAction(snapshot.Map),
            "battle" => ChooseBattleAction(snapshot.Battle),
            "reward" => ChooseRewardAction(snapshot.Reward),
            "event" => ChooseEventAction(snapshot.Event),
            _ => null
        };

        return action ?? ChooseFallbackAction(snapshot);
    }

    private static ActionEnvelope? ChooseMapAction(MapEnvelope? map)
    {
        if (map == null)
        {
            return null;
        }

        var currentRow = map.Rows.FirstOrDefault(row => row.RowIndex == map.CurrentRow);
        if (currentRow == null)
        {
            return null;
        }

        var priority = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase)
        {
            ["NormalBattle"] = 0,
            ["EliteBattle"] = 1,
            ["Event"] = 2,
            ["Rest"] = 3,
            ["Shop"] = 4
        };

        var node = currentRow.Nodes
            .Where(candidate => candidate.CanSelect)
            .OrderBy(candidate => priority.TryGetValue(candidate.NodeType, out var rank) ? rank : 99)
            .ThenBy(candidate => candidate.Column)
            .FirstOrDefault();

        return node == null
            ? null
            : new ActionEnvelope
            {
                Kind = "choose_map_node",
                Column = node.Column
            };
    }

    private static ActionEnvelope? ChooseBattleAction(BattleEnvelope? battle)
    {
        if (battle == null || battle.BattleEnded)
        {
            return null;
        }

        var enemies = battle.Enemies.Where(enemy => enemy.IsAlive).ToList();
        var cards = battle.Hand.Where(card => card.IsPlayable).ToList();

        foreach (var enemy in enemies.OrderBy(enemy => enemy.Hp + enemy.Block))
        {
            foreach (var card in cards.OrderByDescending(EstimateDirectDamage))
            {
                var damage = EstimateDirectDamage(card);
                if (damage <= 0 || !card.RequiresEnemyTarget)
                {
                    continue;
                }

                if (damage >= enemy.Hp + enemy.Block)
                {
                    return new ActionEnvelope
                    {
                        Kind = "play_card",
                        HandIndex = card.HandIndex,
                        TargetEnemyIndex = enemy.EnemyIndex
                    };
                }
            }
        }

        var vulnerable = cards
            .Where(card => card.Description.Contains("Vulnerable", StringComparison.OrdinalIgnoreCase) ||
                           card.Description.Contains("易伤", StringComparison.OrdinalIgnoreCase))
            .OrderBy(card => card.Cost)
            .FirstOrDefault();
        if (vulnerable != null)
        {
            var target = enemies.OrderByDescending(enemy => enemy.Hp + enemy.Block).FirstOrDefault();
            if (target != null)
            {
                return new ActionEnvelope
                {
                    Kind = "play_card",
                    HandIndex = vulnerable.HandIndex,
                    TargetEnemyIndex = target.EnemyIndex
                };
            }
        }

        var bestAttack = cards
            .Where(card => EstimateDirectDamage(card) > 0)
            .OrderByDescending(EstimateDirectDamage)
            .ThenBy(card => card.Cost)
            .FirstOrDefault();
        if (bestAttack != null)
        {
            var target = enemies.OrderBy(enemy => enemy.Hp + enemy.Block).FirstOrDefault();
            if (bestAttack.RequiresEnemyTarget && target != null)
            {
                return new ActionEnvelope
                {
                    Kind = "play_card",
                    HandIndex = bestAttack.HandIndex,
                    TargetEnemyIndex = target.EnemyIndex
                };
            }

            return new ActionEnvelope
            {
                Kind = "play_card",
                HandIndex = bestAttack.HandIndex
            };
        }

        var defense = cards
            .Where(card => card.Description.Contains("Block", StringComparison.OrdinalIgnoreCase) ||
                           card.Description.Contains("格挡", StringComparison.OrdinalIgnoreCase))
            .OrderByDescending(card => card.Cost)
            .FirstOrDefault();
        if (defense != null)
        {
            return new ActionEnvelope
            {
                Kind = "play_card",
                HandIndex = defense.HandIndex
            };
        }

        return new ActionEnvelope { Kind = "end_turn" };
    }

    private static int EstimateDirectDamage(CardEnvelope card)
    {
        var description = card.Description;
        var digitBuffer = string.Empty;
        var best = 0;

        foreach (var ch in description)
        {
            if (char.IsDigit(ch))
            {
                digitBuffer += ch;
                continue;
            }

            if (digitBuffer.Length > 0)
            {
                if (int.TryParse(digitBuffer, out var value))
                {
                    best = Math.Max(best, value);
                }

                digitBuffer = string.Empty;
            }
        }

        if (digitBuffer.Length > 0 && int.TryParse(digitBuffer, out var trailing))
        {
            best = Math.Max(best, trailing);
        }

        return best;
    }

    private static ActionEnvelope? ChooseRewardAction(RewardEnvelope? reward)
    {
        if (reward == null)
        {
            return null;
        }

        if (string.Equals(reward.Mode, "card_pack", StringComparison.OrdinalIgnoreCase))
        {
            var bestCard = reward.CardOptions
                .OrderByDescending(option => ScoreRewardCard(option))
                .ThenBy(option => option.OptionIndex)
                .FirstOrDefault();

            return bestCard == null
                ? new ActionEnvelope { Kind = "skip_reward" }
                : new ActionEnvelope
                {
                    Kind = "choose_reward_card",
                    OptionIndex = bestCard.OptionIndex
                };
        }

        var preference = reward.RewardTypes.Contains("relic", StringComparer.OrdinalIgnoreCase)
            ? "relic"
            : reward.RewardTypes.Contains("card_pack", StringComparer.OrdinalIgnoreCase)
                ? "card_pack"
                : reward.RewardTypes.Contains("potion", StringComparer.OrdinalIgnoreCase)
                    ? "potion"
                    : "random";

        return new ActionEnvelope
        {
            Kind = "choose_reward_type",
            RewardType = preference
        };
    }

    private static int ScoreRewardCard(RewardOptionEnvelope option)
    {
        var text = $"{option.Name} {option.Description}";
        var score = 0;
        if (text.Contains("Strength", StringComparison.OrdinalIgnoreCase) || text.Contains("力量", StringComparison.OrdinalIgnoreCase))
        {
            score += 40;
        }
        if (text.Contains("Energy", StringComparison.OrdinalIgnoreCase) || text.Contains("能量", StringComparison.OrdinalIgnoreCase))
        {
            score += 35;
        }
        if (text.Contains("Vulnerable", StringComparison.OrdinalIgnoreCase) || text.Contains("易伤", StringComparison.OrdinalIgnoreCase))
        {
            score += 20;
        }
        score += text.Count(char.IsDigit);
        return score;
    }

    private static ActionEnvelope? ChooseEventAction(EventEnvelope? eventState)
    {
        if (eventState == null)
        {
            return null;
        }

        return new ActionEnvelope
        {
            Kind = "choose_event_option",
            OptionIndex = 0
        };
    }

    private static ActionEnvelope? ChooseFallbackAction(SnapshotEnvelope snapshot)
    {
        var legalAction = snapshot.LegalActions.FirstOrDefault();
        if (legalAction == null)
        {
            return null;
        }

        var parameters = legalAction.Parameters;
        var action = new ActionEnvelope
        {
            Kind = legalAction.Kind
        };

        if (parameters.TryGetPropertyValue("column", out var columnNode) && columnNode != null)
        {
            action.Column = columnNode.GetValue<int>();
        }
        if (parameters.TryGetPropertyValue("handIndex", out var handIndexNode) && handIndexNode != null)
        {
            action.HandIndex = handIndexNode.GetValue<int>();
        }
        if (parameters.TryGetPropertyValue("cardId", out var cardIdNode) && cardIdNode != null)
        {
            action.CardId = cardIdNode.GetValue<string>();
        }
        if (parameters.TryGetPropertyValue("rewardType", out var rewardTypeNode) && rewardTypeNode != null)
        {
            action.RewardType = rewardTypeNode.GetValue<string>();
        }
        if (parameters.TryGetPropertyValue("optionIndex", out var optionIndexNode) && optionIndexNode != null)
        {
            action.OptionIndex = optionIndexNode.GetValue<int>();
        }
        if (parameters.TryGetPropertyValue("targetEnemyIndices", out var targetEnemyIndicesNode)
            && targetEnemyIndicesNode is System.Text.Json.Nodes.JsonArray targets
            && targets.Count > 0
            && targets[0] != null)
        {
            action.TargetEnemyIndex = targets[0]!.GetValue<int>();
        }

        return action;
    }
}
